akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -304,6 +358,36 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
         settingsAppliedOk
       }
 
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case Keepalive.Tick =>
+          pingState match {
+            case OptionVal.Some(state) =>
+              // don't do anything unless there are active streams
+              debug(s"keepalive tick, activeStreamCount: ${activeStreamCount()}, ticksWithoutData: ${state.ticksWithoutData}, pingsWithoutData: ${state.pingsWithoutData}, pingEveryN: ${state.pingEveryNTickWithoutData}")
+              if (activeStreamCount() > 0) {
+                state.ticksWithoutData += 1
+                val now = System.nanoTime()
+                if (state.lastPingTimestamp > 0L && (state.lastPingTimestamp - now) > state.maxKeepaliveTimeoutNanos) {
```

Comment:
`now - state.lastPingTimestamp`  instead of `state.lastPingTimestamp - now`?

Should `maxKeepaliveTimeoutNanos` be checked when receiving a Ping(ACK)? As it is here, it looks like certain settings will make this code branch useless (if pings are sent more often than the configured `maxKeepaliveTimeoutNanos`), no?

A hacky way to check the ping ACK was received faster than `maxKeepaliveTimeoutNanos` would be to put the value of `now` as the ping payload.

Reply:
ups. yes, `now - lastPing` for sure.

`timeout` must be lower than `interval`, will be verified in settings loading before done.

I think it is better to not trust the remote side with data we use for something (I think the spec mandates that it is the same ping content that is in the pong though, so maybe we should verify that)